R/functions-Spectrum.R

```diff
@@ -568,3 +587,462 @@ validSpectrum <- function(object) {
+#'
+#' ## Use the weighted average considering the adjacent mz
+#' mzs_1 <- descendPeak(mz = mzs, peakIdx = peak_idx, intensity = ints)
+#' points(mzs_1[, 1], mzs_1[, 2], col = "red", type = "h")
+#'
+#' ## Values considered for the first peak:
+#' pk_1 <- c(1, 2, 3, 4, 5)
+#' mzs_1[1, 1] == weighted.mean(mzs[pk_1], ints[pk_1])
+#'
+#' ## Second example 
+#' ints <- c(5, 3, 2, 3, 1, 2, 4, 6, 8, 11, 4, 7, 5, 2, 1, 0, 1, 4, 1, 1, 1, 0)
+#' mzs <- 1:length(ints)
+#'
+#' plot(mzs, ints, type = "h")
+#' peak_idx <- 10
+#' points(mzs[peak_idx], ints[peak_idx], pch = 16)
+#'
+#' mzs_2 <- descendPeak(mz = mzs, peakIdx = peak_idx, intensity = ints,
+#'     signalPercentage = 0)
+#' points(mzs_2[, 1], mzs_2[, 2], col = "red", type = "h")
+#'
+#' ## Points considered:
+#' pk_idx <- c(5, 6, 7, 8, 9, 10, 11)
+#' mzs_2[1, 1] == weighted.mean(mzs[pk_idx], ints[pk_idx])
+#'
+#' ## Stopping at two increasing values
+#' mzs_2 <- descendPeak(mz = mzs, intensity = ints, peakIdx = peak_idx,
+#'     signalPercentage = 0, stopAtTwo = TRUE)
+#' points(mzs_2[, 1], mzs_2[, 2], col = "blue", type = "h")
+#' pk_idx <- c(3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
+#' mzs_2[1, 1] == weighted.mean(mzs[pk_idx], ints[pk_idx])
+#'
+#' 
+#' ## Include "missing" measurements.
+#' ints <- ints[-9]
+#' mzs <- mzs[-9]
+#'
+#' plot(mzs, ints, type = "h")
+#' peak_idx <- 9
+#'
+#' points(mzs[peak_idx], ints[peak_idx], pch = 16)
+#' mzs_2 <- descendPeak(mz = mzs, peakIdx = peak_idx, intensity = ints,
+#'     signalPercentage = 0)
+#' points(mzs_2[, 1], mzs_2[, 2], col = "red", type = "h")
+descendPeak <- function(mz, intensity, peakIdx = NULL, signalPercentage = 33,
+                        stopAtTwo = FALSE, ...) {
+    if (length(mz) != length(intensity))
+        stop("lengths of 'mz' and 'intensity' have to match")
+    if (length(peakIdx) == 0)
+        return(cbind(mz = mz[peakIdx], intensity = intensity[peakIdx]))
+    len <- length(mz)
+    signalPercentage = signalPercentage / 100
+    max_k <- 30                        # max peak region to consider
+    ## Define the index of values to include.
+    do.call(rbind, lapply(peakIdx, function(z) {
+        ## Define the region of interest in which we will search for signal
+        ## larger than the threshold. We restrict to a max region +- max_k
+        ## data points large.
+        ## 1) Ensure that the region we consider is symmetric around the
+        ##    peak - important at the edges.
+        from_to <- c(max(1, z - max_k), min(len, z + max_k))
+        half_width <- min(abs(from_to - z))
+        ## Descend to the right
+        to_idx <- z + half_width
+        ## 2) Restrict the region to the area with monotonically decreasing
+        ##    signal (from the apex).
+        tmp_idx <- z:to_idx
+        sign_change <- diff(intensity[tmp_idx]) >= 0
+        ## Allow a single increasing measurement and stop descending if
+        ## there are two consecutive increasing values.
+        if (stopAtTwo)
```

Comment:
Are two consecutive increasing values always enough (I guess it depends on the smoothing)?

Reply:
After smoothing it shouldn't even be necessary to allow 2 consecutive increasing values. And I guess for most platforms it would not be necessary also for un-smoothed data.